### ISSUE TITLE
Zero QUAL for reference call

### ISSUE THREAD

**Issue by alexjironkin** [samtools/samtools#566]
Some of the reference positions in VCF file after mpileup + bcftools come back with 0 QUAL. This behaviour is different between 1.1 and 1.3. 

v1.1
```
gi|16445344|gb|AE006468.1|      59      .       A       .       999     .       DP=70;MQSB=1;MQ0F=0;AN=2;DP4=36,23,0,0;MQ=60    GT:DP:DV:SP:DP4:DPR     0/0:59:0:0:36,23,0,0:59
```

v1.3
```
gi|16445344|gb|AE006468.1|      59      .       A       .       0       .       DP=70;ADF=35;ADR=18;AD=53;MQSB=1;MQ0F=0;AN=2;DP4=35,18,0,0;MQ=60        GT:DP:SP:ADF:ADR:AD     0/0:53:0:35:18:53
```

Looks like ```DP``` value has changed as well. It doesn't feel like this was intended, but is this expected behaviour?

**Comment by alexjironkin** [comment]
Going through the VCF  spec:

```QUAL - quality: Phred-scaled quality score for the assertion made in ALT. i.e. −10log10 prob(call in ALT is
wrong). If ALT is ‘.’ (no variant) then this is −10log10 prob(variant), and if ALT is not ‘.’ this is −10log10
prob(no variant). If unknown, the missing value should be specified. (Numeric)```

if QUAL is 0 with ALT = "." then Prob(var) = 10^(0/-10) = 1, suggesting that should be a variant.

**Comment by alexjironkin** [comment]
This seems to be an issue with bcftools, not samtools.

**Comment by pd3** [comment]
Raw mpileup output does not compute qualitites, these are filled by `bcftools call`. We can, however, set the QUALs to missing values `.` to help avoid the confusion.

**Comment by alexjironkin** [comment]
I'm not sure that would help. I asked what's the state of that issue in bcftools thread. It seems to have been closed, but we are seeing this output in the latest releases and it doesn't seem to be correct.

**Comment by pd3** [comment]
Ah, you are saying that you are observing QUAL=0 in `bcftools call` output. Sorry, my mistake! 
Thanks for letting us know about the problem, this should fix it https://github.com/pd3/bcftools/commit/d937dc11d7ebf4fa21ec71e2b1a9049aec6ea65d.

**Comment by alexjironkin** [comment]
Great. I just pulled your dev branch and tested it with my mock dataset and it seems to work now. No QUAL=0 when there were before. 

As an aside, QUAL score topped at 999? What was the rational behind that? Can that cap be moved?

**Comment by alexjironkin** [comment]
I presume QUAL=0 fix is going to be available in the next release?

Thanks

**Comment by pd3** [comment]
Indeed. 

Larger values have no added benefit, the precision down to 1e-100 should surely suffice?!

**Comment by alexjironkin** [comment]
We were considering something akin to [GATK's QD field](https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_annotator_QualByDepth.php). However, if QUAL is capped then for large DP values it would produce wrong result. Intuitively, it seems to make sense to include depth into quality calculation.

**Comment by pd3** [comment]
How about you test first on a locally modified version, and if it proves useful, we consider leaving the values untrimmed?

For this, just comment out these lines
https://github.com/pd3/bcftools/blob/develop/mcall.c#L1539-L1540

**Comment by alexjironkin** [comment]
Yeah, I noticed them in the commit diff. I'll try that, and will raise a request if it can be useful.

Thanks

**Comment by talshmaya** [comment]
Hi, we applied that patch ^ (https://github.com/samtools/bcftools/pull/449) and it did fix the issue with qual of 0 for high depth calls, but now it assigns a qual score of 999 for bases with depth of 0 and/or high qual scores for very low depth. Examples:

`gi|448814763|ref|NC_000962.3|   177058  .       T       .       999     .       DP=0;MQ0F=0;AN=0;DP4=0,0,0,0;MQ=.       GT      .`


`gi|448814763|ref|NC_000962.3|   103929  .       G       .       68      .       DP=1;MQ0F=0;AN=1;DP4=0,1,0,0;MQ=40      GT      0
gi|448814763|ref|NC_000962.3|   103930  .       G       C       8.11273 .       DP=1;SGB=-0.379885;MQ0F=0;AC=1;AN=1;DP4=0,0,0,1;MQ=40   GT:PL   1:36,0
gi|448814763|ref|NC_000962.3|   103931  .       C       .       68      .       DP=1;MQ0F=0;AN=1;DP4=0,1,0,0;MQ=40      GT      0
`

Was that issue addressed in the newest build? 

Thanks!

**Comment by pd3** [comment]
Hi, 
this is a tricky question. I believe it is not a bug but a side effect of the model: there is a prior on the reference genotype, so unless there is a strong evidence for a non-reference allele, the quality of a 0/0 call will be high. The concrete value will depend on mapping and base qualities, number of samples etc.

The QUAL=999 with DP=0 site is interesting. Could you please extract this line from the mpileup output and send it to me take a closer look? There is an email address on my profile page.
Thanks

**Comment by pd3** [comment]
I am reopening the issue, it was closed by mistake. This should be improved, although it's not clear yet how.
